File: app/src/main/java/de/xikolo/jobs/UpdateAnnouncementVisitedJob.java
package de.xikolo.jobs;

import android.util.Log;

import com.birbit.android.jobqueue.Params;

import de.xikolo.config.Config;
import de.xikolo.jobs.base.BaseJob;
import de.xikolo.jobs.base.JobCallback;
import de.xikolo.managers.UserManager;
import de.xikolo.models.Announcement;
import de.xikolo.models.base.Sync;
import de.xikolo.network.ApiService;
import io.realm.Realm;
import retrofit2.Response;

public class UpdateAnnouncementVisitedJob extends BaseJob {

    public static final String TAG = UpdateAnnouncementVisitedJob.class.getSimpleName();

    private String announcementId;

    public UpdateAnnouncementVisitedJob(String announcementId) {
        super(new Params(PRIORITY_LOW).requireNetwork().persist(), null);

        this.announcementId = announcementId;
    }

    @Override
    public void onAdded() {
        if (Config.DEBUG) Log.i(TAG, TAG + " added | announcement.id " + announcementId);

        Sync.Update.with(Announcement.class, announcementId)
                .setBeforeCommitCallback(new Sync.BeforeCommitCallback<Announcement>() {
                    @Override
                    public void beforeCommit(Realm realm, Announcement model) {
                        model.visited = true;
                    }
                })
                .run();
    }

    @Override
    public void onRun() throws Throwable {
        if (UserManager.isAuthorized()) {
            Announcement.JsonModel model = new Announcement.JsonModel();
            model.setId(announcementId);
            model.visited = true;

            final Response<Announcement.JsonModel> response = ApiService.getInstance().updateAnnouncement(
                    UserManager.getTokenAsHeader(),
                    announcementId,
                    model
            ).execute();

            if (response.isSuccessful()) {
                if (Config.DEBUG) Log.i(TAG, "Announcement visit successfully updated");

                if (callback != null) callback.success();
            } else {
                if (Config.DEBUG) Log.e(TAG, "Error while updating announcement visit");
                if (callback != null) callback.error(JobCallback.ErrorCode.ERROR);
            }
        } else {
            if (callback != null) callback.error(JobCallback.ErrorCode.NO_AUTH);
        }
    }

}

File: app/src/main/java/de/xikolo/jobs/UpdateAnnouncementVisitedJob.kt
package de.xikolo.jobs

import android.util.Log
import com.evernote.android.job.JobRequest
import com.evernote.android.job.util.support.PersistableBundleCompat
import de.xikolo.config.Config
import de.xikolo.jobs.base.ScheduledJob
import de.xikolo.models.Announcement
import de.xikolo.models.base.Local
import de.xikolo.models.base.Sync
import de.xikolo.network.ApiService

class UpdateAnnouncementVisitedJob : ScheduledJob(Precondition.AUTH) {

    companion object {
        val TAG: String = UpdateAnnouncementVisitedJob::class.java.simpleName

        @JvmStatic
        fun schedule(announcementId: String): Int {
            if (Config.DEBUG) Log.i(TAG, TAG + " scheduled | announcement.id " + announcementId)

            Local.Update.with(Announcement::class.java, announcementId)
                    .setBeforeCommitCallback { _, model -> model.visited = true }
                    .run()

            val extras = PersistableBundleCompat()
            extras.putString("ann_id", announcementId)

            return JobRequest.Builder(UpdateAnnouncementVisitedJob.TAG)
                    .setExecutionWindow(1, 30_000)
                    .setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
                    .setRequirementsEnforced(true)
                    .setExtras(extras)
                    .build()
                    .schedule()
        }
    }

    override fun onRun(params: Params): Result {
        val model = Announcement.JsonModel()
        model.id = params.extras.getString("ann_id", null)
        model.visited = true

        val response = ApiService.getInstance().updateAnnouncement(model.id, model).execute()

        return if (response.isSuccessful) {
            if (Config.DEBUG) Log.i(TAG, "Announcement visit successfully updated")

            Sync.Data.with(Announcement::class.java, response.body())
                    .saveOnly()
                    .run()

            Result.SUCCESS
        } else {
            if (Config.DEBUG) Log.e(TAG, "Error while updating announcement visit")
            Result.FAILURE
        }
    }

}
